Index: config/SundialsSuperLUMT.cmake
===================================================================
--- config/SundialsSuperLUMT.cmake	(revision 13763)
+++ config/SundialsSuperLUMT.cmake	(working copy)
@@ -64,6 +64,9 @@
   if(LTEST_OK)
     message(STATUS "Checking if SUPERLUMT works... OK")
     set(SUPERLUMT_FOUND TRUE)
+    if(MSVC)
+        ADD_DEFINITIONS(/D __OPENMP)
+    endif()
   else(LTEST_OK)
     message(STATUS "Checking if SUPERLUMT works... FAILED")
   endif(LTEST_OK)
